File: Downloads/ai-plans-strengths_and_vulnernabilities/ai-plans-strengths_and_vulnernabilities/server/get_arxiv_data.js
const axios = require('axios');
const fs = require('fs');
const log = require('./log.js')
const { parseString } = require('xml2js');



function extractArxivId( url ) {
    // Regular expressions for matching ArXiv identifiers
    const newScheme = /\d{4}\.\d+(v\d+)?/;
    const oldScheme = /([\w\.-]+\/\d{4}\d+)/;
    
    const match = url.match( newScheme ) || url.match( oldScheme );
    if ( match ) {
        return match[0];
    }
    return null;
}

async function getArxivPaper( url ) {

    const articleId = extractArxivId( url );
    if ( !articleId ) {
    	log('flag', 'failed to extract arxiv id', articleId, url )
        return null;
    }

    try {

    	const try_url = `http://export.arxiv.org/api/query?id_list=${articleId}`

    	log('flag', 'fetching XML data: ', try_url )

        const response = await axios.get( try_url );

        const paperData = await parseXML( response.data )

        paperData.extracted_id = articleId

        log('flag', 'axios XML parse res: ',  paperData )

        // response.data.feed.entry[0];

        return paperData;

    } catch (error) {
    	log('flag', error )
        return null;
    }
}

function parseXML(xmlData) {
    return new Promise((resolve, reject) => {
        parseString(xmlData, (err, result) => {
            if (err) {
                reject(err);
            } else {
                resolve(result.feed.entry[0]);
            }
        });
    });
}

function ensurePathExists(path) {
    if (!fs.existsSync(path)) {
        fs.mkdirSync(path);
    }
}

async function getArxivData( url, args ) {

	try{

	    // const arguments = process.argv.slice(2);
    
	    if (args.length === 0) {
	        throw new Error('Received no arguments.');
	    }
	    
	    // const url = args[0];

	    log('flag', 'given args: ', url, args )
	    
	    const paper = await getArxivPaper( url );
	    if (!paper) {
	        return {
	        	success: false,
	        	msg: 'Article not found'
	        }
	    }

	    const data = {};

	    data.title = paper.title[0];

	    data.authors = paper.author.map(author => author.name);

	    data.summary = paper.summary;

	    // log('flag', 'what is paper ', paper )
	    // log('flag', 'can we do clearner URL... ', paper.id[0] )

	    const paper_slug = get_paper_slug( paper.id[0] )
	    if( typeof paper_slug !== 'string' || paper_slug.length < 5 ) return {
	    	success: false,
	    	msg: 'invalid paper URL generated',
	    }

	    const pdfPath = `./fs/arxiv/${ paper_slug }.pdf`;
	    ensurePathExists('./fs/arxiv/');

		if( fs.existsSync( pdfPath ) ){

			log('flag', 'paper already exists: ', pdfPath )

		}else{

		    const pdf_link = paper.link[1]?.['$']?.href  // what is that ??

		    // log('flag', 'paper link data: ', paper.link[0] )
		    // log('flag', 'paper link data: ', paper.link[1] )

		    const response = await axios({
		        method: 'get',
		        url: pdf_link,
		        responseType: 'stream'
		    })

	        response.data.pipe( fs.createWriteStream( pdfPath ) );
		}

		data.pdf_path = pdfPath;
		data.embed_ref = pdfPath.replace('./fs/arxiv/', '').replace('.pdf', '')
	    
	    return {
	    	success: true,
	    	data: data
	    };

	}catch( err ){
		log('flag', 'arxiv err', err )

	}

}

const get_paper_slug = url => {
	if( typeof url !== 'string' ) return log('flag', 'invalid paper url', url )
	return url.replace(/https?:?\/?\/?/, '').replace(/\//g, '_')
}

module.exports = getArxivData

File: Downloads/ai-plans-strengths_and_vulnernabilities/ai-plans-strengths_and_vulnernabilities/server/py/get_arxiv_data.py
# Depends on the `arxiv` module, which can be installed with
#     pip install arxiv
# More info on the API:
# - https://pypi.org/project/arxiv/
# - https://info.arxiv.org/help/api/index.html

import arxiv
import json
import re
import os
import sys

def extract_arxiv_id(url):
    # https://info.arxiv.org/help/arxiv_identifier.html
    
    # Identifier scheme since 1 April 2007
    new_scheme = r'\d{4}\.\d+(v\d+)?'
    # Identifiers up to March 2007
    old_scheme = r'([\w\.-]+/\d{4}\d+)'
    
    # Check for matches against the two schemes
    for pattern in (new_scheme, old_scheme):
        match = re.search(pattern, url)
        if match:
            article_id = match.group(0)
            return article_id
    
    # If no match is found, return None
    return None

def get_arxiv_paper(url):
    # https://pypi.org/project/arxiv/
    
    # Extract the article id
    article_id = extract_arxiv_id(url)
    # Stop if we can't find a valid id
    if article_id is None:
        return None
    
    # Let's see if the id matches an actual article
    try:
        # Use API to find the article
        search = arxiv.Search(id_list=[article_id])
        paper = next(search.results())
        return paper
    except:
        # Article doesn't exist, or API is down, or something else.
        return None

def ensure_path_exists(path):
    if not os.path.exists(path):
        os.mkdir(path)

def save_pdf(paper):
	# Save the paper's pdf to ./papers/
	
	# This directory is relative to the project root
    path = './fs/py/'
    ensure_path_exists(path)
    pdf_path = paper.download_pdf(path)
    return pdf_path

def get_arxiv_data():
	# Gets the article's title, authors, summary and saves its pdf,
	# converts that data to json (as for the pdf, only its save path
	# goes in the json), and prints the json string.
	
    arguments = sys.argv[1:]
    
    if len(arguments) == 0:
    	raise TypeError('Received no arguments.')
    
    url = arguments[0]

    paper = get_arxiv_paper(url)

    # Get data from the paper
    data = {}

    data['title'] = paper.title

    data['authors'] = [
        str(author) for author in paper.authors
    ]

    data['summary'] = paper.summary

    pdf_path = save_pdf(paper)
    data['pdf_path'] = pdf_path
    
    return data

def main():
	try:
		data = get_arxiv_data()
	except Exception as error:
		data = {
			'error' : {
				'type': type(error).__name__,
				'description': str(error)
			}
		}
	
	print(json.dumps(data))

main()

